src/dump.complex.js

```diff
@@ -0,0 +1,100 @@
+'use strict'
+
+const dumpBool = require('./dump.bool')
+const dumpFloat = require('./dump.float')
+const dumpInt = require('./dump.int')
+const dumpNull = require('./dump.null')
+const dumpString = require('./dump.string')
+const dumpSymbol = require('./dump.symbol')
+const dumpSymbolLink = require('./dump.symbol.link')
+
+/**
+ * @param {string} symbol
+ * @param {Array<string>} symcache
+ * @returns {Buffer}
+ */
+const dumpSymbolOrLink = (symbol, symcache) => {
+  const pos = symcache.indexOf(symbol)
+  if (pos === -1) {
+    symcache.push(symbol)
+    return dumpSymbol(symbol)
+  }
+
+  return dumpSymbolLink(pos)
+}
+
+/**
+ * @param {Array} arr
+ * @param {Array<string>} symcache
+ * @returns {Buffer}
+ */
+const dumpArray = (arr, symcache) => {
+  const head = Buffer.from('[', 'utf8')
+  const len = dumpInt(arr.length).slice(1)
+  const data = []
+
+  for (const val of arr) {
+    data.push(encodeType(val, symcache))
+  }
+
+  return Buffer.concat([
+    head,
+    len,
+    ...data
+  ])
+}
+
+/**
+ * @param {Object} obj
+ * @param {Array<string>} symcache
+ * @returns {Buffer}
+ */
+const dumpHash = (obj, symcache) => {
+  const head = Buffer.from('{', 'utf8')
+  const len = dumpInt(Object.keys(obj).length).slice(1)
+
+  const data = []
+  for (const key in obj) {
+    data.push(dumpSymbolOrLink(key, symcache), encodeType(obj[key], symcache))
+  }
+
+  return Buffer.concat([
+    head,
+    len,
+    ...data
+  ])
+}
+
+/**
+ * @param {any} val
+ * @param {Array<string>} symcache
+ * @returns {Buffer}
+ */
+const encodeType = (val, symcache) => { // see http://jakegoulding.com/blog/2013/01/15/a-little-dip-into-rubys-marshal-format
+  let symbol
+
+  switch (typeof val) {
+    case 'string': return dumpString(val)
+    case 'boolean': return dumpBool(val)
+    case 'number':
+      if (Number.isInteger(val)) return dumpInt(val)
```

Comment:
edge case?

```js
Number.isInteger(1.0)
true
```

Reply:
```
2.7.2 :001 > Marshal.dump(1.0)
 => "\x04\bf\x061"
2.7.2 :002 > Marshal.dump(1)
 => "\x04\bi\x06"
 ```

Reply:
this seems as something nodejs always treats that way:
```
repl
> 1.0
1
> (1.0).toString()
'1'
```
so we can't actually detect this case in nodejs, but it doesn't look a big issue to me since ruby would work with each format

Reply:
at some point you'll run into issues with that in certain edge cases. example:

```
2.7.2 :010 > 1.to_s
 => "1"
2.7.2 :011 > 1.0.to_s
 => "1.0"
```

some sloppy code using this behaviour:

```ruby
foo = 1.0

def is_float(n)
  n.to_s.end_with?('.0')
end 

is_float(foo)
```

Reply:
we could solve with a special shim for it:

```
if (Number.isInteger(val)) return dumpInt(val)

function dumpInt (val) {
  if (('' + val).contains('.0') {
    // manually encode here
  }
}
```

Reply:
agreed to add info in docs since we can't determine 1.0 as 1.0 in nodejs